core/services/vrf/v2/listener_v2_log_listener_test.go

```diff
@@ -0,0 +1,1020 @@
+		require.NoError(t, err1)
+		_, err1 = th.Emitter.EmitLog2(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		th.Client.Commit()
+	}
+
+	// Blocks till now: 2 (in SetupTH) + 2 (empty blocks) + 5 (EmitLog blocks) = 9
+
+	// Calling Start() after RegisterFilter() simulates a node restart after job creation, should reload Filter from db.
+	require.NoError(t, th.LogPoller.Start(testutils.Context(t)))
+
+	// The poller starts on a new chain at latest-finality (finalityDepth + 5 in this case),
+	// Replaying from block 4 should guarantee we have block 4 immediately.  (We will also get
+	// block 3 once the backup poller runs, since it always starts 100 blocks behind.)
+	require.NoError(t, th.LogPoller.Replay(testutils.Context(t), 4))
+
+	// We should immediately have at least logs 4-7
+	logs, err := th.LogPoller.Logs(4, 7, emitterABI.Events["Log1"].ID, th.EmitterAddress,
+		pg.WithParentCtx(testutils.Context(t)))
+	require.NoError(t, err)
+	require.Equal(t, 3, len(logs))
+
+	lastProcessedBlock, err := th.Listener.initializeLastProcessedBlock(th.Ctx)
+	require.Nil(t, err)
+	require.Equal(t, int64(6), lastProcessedBlock)
+}
+
+func TestInitProcessedBlock_NoUnfulfilledVRFReqs(t *testing.T) {
+	t.Parallel()
+
+	finalityDepth := int64(3)
+	th := setupVRFLogPollerListenerTH(t, false, finalityDepth, 3, 2, 1000, func(mockChain *evmmocks.Chain, curTH *vrfLogPollerListenerTH) {
+		mockChain.On("ID").Return(curTH.ChainID)
+		mockChain.On("LogPoller").Return(curTH.LogPoller)
+	})
+
+	// Block 3 to finalityDepth. Ensure we have finality number of blocks
+	for i := 1; i < int(finalityDepth); i++ {
+		th.Client.Commit()
+	}
+
+	// Create VRF request block and a fulfillment block
+	keyHash := [32]byte(th.Listener.job.VRFSpec.PublicKey.MustHash().Bytes())
+	preSeed := big.NewInt(105)
+	subID := uint64(1)
+	reqID := big.NewInt(1)
+	_, err2 := th.VRFLogEmitter.EmitRandomWordsRequested(th.Owner,
+		keyHash, reqID, preSeed, subID, 10, 10000, 2, th.Owner.From)
+	require.NoError(t, err2)
+	th.Client.Commit()
+	_, err2 = th.VRFLogEmitter.EmitRandomWordsFulfilled(th.Owner, reqID, preSeed, big.NewInt(10), true)
+	require.NoError(t, err2)
+	th.Client.Commit()
+
+	// Emit some logs in blocks to make the VRF req and fulfillment older than finalityDepth from latestBlock
+	numBlocks := 5
+	for i := 0; i < numBlocks; i++ {
+		_, err1 := th.Emitter.EmitLog1(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		_, err1 = th.Emitter.EmitLog2(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		th.Client.Commit()
+	}
+
+	// Blocks till now: 2 (in SetupTH) + 2 (empty blocks) + 2 (VRF req/resp block) + 5 (EmitLog blocks) = 11
+
+	// Calling Start() after RegisterFilter() simulates a node restart after job creation, should reload Filter from db.
+	require.NoError(t, th.LogPoller.Start(th.Ctx))
+
+	latestBlock := finalityDepth + int64(numBlocks) + 1
```

Comment:
Yes, request is in block 5 and fulfillment is in block 6. One replay outside seems to be needed for this test to work, otherwise there is some issue with DB connection being closed. So, I replay from the latest block for this and then check if `initializeLastProcessedBlock` internally replay's from block 5 to process the request.

Reply:
yeah a replay is probably needed so that log poller has a latest block. Otherwise, `initializeLastProcessedBlock` will fail because of empty rows for `latestHead()`

Would it be better to just set `latestBlock := uint64(1)` to cache the latest block? Then, `initializeLastProcessedBlock` will still find the correct last processed block and test should still succeed. I wasn't able to figure out what `finalityDepth + int64(numBlocks) + 1` means

Reply:
same comment for other `initializeLastProcessedBlock` tests

Reply:
ah ok. If we set `latestBlock := uint64(1)`, then we would externally replay from block 1 onwards. But we'd like to test if `initializeLastProcessedBlock` is able to replay from the earliestUnprocessedBlock right